src/js/common/schemaform/formState.js

```diff
@@ -0,0 +1,378 @@
+
+    if (newItems !== schema.items) {
+      return _.set('items', newItems, schema);
+    }
+  }
+
+  return schema;
+}
+
+/**
+ * This function updates an array schema to use the array of
+ * item schema format and keeps that array in sync with the
+ * data in that array in the form data.
+ *
+ * This allows us to have conditional fields for each array item,
+ * because our conditional field implementation depends on modifying
+ * schemas
+ *
+ * @param {Object} schema The current JSON Schema object
+ * @param {any} fieldData The data associated with the current schema
+ */
+export function updateItemsSchema(schema, fieldData = null) {
+  if (schema.type === 'array') {
+    let newSchema = schema;
+
+    // This happens the first time this function is called when
+    // generating the form
+    if (!Array.isArray(schema.items)) {
+      newSchema = _.assign(schema, {
+        items: [],
+        additionalItems: schema.items
+      });
+    }
+
+    if (!fieldData) {
+      // If there's no data, the list of schemas should be empty
+      return _.set('items', [], newSchema);
+    } else if (fieldData.length > newSchema.items.length) {
+      // Here we're filling in the items array to make it the same
+      // length as the array of form data. This happens when you add
+      // another record on the form, mainly.
+      const fillIn = Array(fieldData.length - newSchema.items.length)
+        .fill(newSchema.additionalItems);
+      return _.set('items', newSchema.items.concat(fillIn), newSchema);
+    } else if (fieldData.length < newSchema.items.length) {
+      // If someone removed a record we're removing the last schema item
+      // This may not be the actual removed schema, but the schemas will
+      // always be updated in the next step
+      return _.set('items', _.dropRight(1, newSchema.items), newSchema);
+    }
+
+    return newSchema;
+  }
+
+  if (schema.type === 'object') {
+    const newSchema = Object.keys(schema.properties).reduce((current, next) => {
+      const nextProp = updateItemsSchema(schema.properties[next], fieldData ? fieldData[next] : null);
+
+      if (current.properties[next] !== nextProp) {
+        return _.set(['properties', next], nextProp, current);
+      }
+
+      return current;
+    }, schema);
+
+    return newSchema;
+  }
+
+  return schema;
+}
+
+/**
+ * This is the main sequence of updates that happens when data is changed
+ * on a form. Most updates are applied to the schema, except that the data
+ * is updated to remove newly hidden data
+ *
+ * @param {Object} schema The current JSON Schema
+ * @param {Object} uiSchema The current UI Schema (does not change)
+ * @param {Object} formData Flattened data for the entire form
+ * @param {Object} pageData The data for the current page
+ * @param {Object} state The Redux state object, which contains schema and data for
+ *   all pages
+ */
+export function updateSchemaData(schema, uiSchema, formData, pageData, state) {
```

Comment:
I don't have a better name, but this one is kind of misleading, isn't it? Mostly it updates the schema...but also updates the data as well.

`updateSchemaAndData` seems a bit much, but it's accurate enough...